src/pages/workspace/AccessOrNotFoundWrapper.tsx

```diff
@@ -0,0 +1,109 @@
+/* eslint-disable rulesdir/no-negated-variables */
+import React, {useEffect} from 'react';
+import type {OnyxEntry} from 'react-native-onyx';
+import {withOnyx} from 'react-native-onyx';
+import FullPageNotFoundView from '@components/BlockingViews/FullPageNotFoundView';
+import FullscreenLoadingIndicator from '@components/FullscreenLoadingIndicator';
+import Navigation from '@libs/Navigation/Navigation';
+import * as PolicyUtils from '@libs/PolicyUtils';
+import NotFoundPage from '@pages/ErrorPage/NotFoundPage';
+import * as Policy from '@userActions/Policy';
+import ONYXKEYS from '@src/ONYXKEYS';
+import ROUTES from '@src/ROUTES';
+import type * as OnyxTypes from '@src/types/onyx';
+import type {PolicyFeatureName} from '@src/types/onyx/Policy';
+import callOrReturn from '@src/types/utils/callOrReturn';
+import {isEmptyObject} from '@src/types/utils/EmptyObject';
+
+const POLICY_ACCESS_VARIANTS = {
+    PAID: (policy: OnyxEntry<OnyxTypes.Policy>) => !PolicyUtils.isPaidGroupPolicy(policy) || !policy?.isPolicyExpenseChatEnabled,
+    ADMIN: (policy: OnyxEntry<OnyxTypes.Policy>) => !PolicyUtils.isPolicyAdmin(policy),
+} as const satisfies Record<string, (policy: OnyxTypes.Policy) => boolean>;
+
+type PolicyAccessVariant = keyof typeof POLICY_ACCESS_VARIANTS;
+
+type AccessOrNotFoundWrapperOnyxProps = {
+    /** The report currently being looked at */
+    policy: OnyxEntry<OnyxTypes.Policy>;
+
+    /** Indicated whether the report data is loading */
+    isLoadingReportData: OnyxEntry<boolean>;
+};
+
+type AccessOrNotFoundWrapperProps = AccessOrNotFoundWrapperOnyxProps & {
+    /** The children to render */
+    children: ((props: AccessOrNotFoundWrapperOnyxProps) => React.ReactNode) | React.ReactNode;
+
+    /** The report currently being looked at */
+    policyID: string;
+
+    /** Defines which types of access should be verified */
+    accessVariants?: PolicyAccessVariant[];
+
+    /** The current feature name that the user tries to get access to */
+    featureName?: PolicyFeatureName;
+};
+
+type PageNotFoundFallackProps = Pick<AccessOrNotFoundWrapperProps, 'policyID'> & {showFullScreenFallback: boolean};
+
+function PageNotFoundFallback({policyID, showFullScreenFallback}: PageNotFoundFallackProps) {
+    return showFullScreenFallback ? (
+        <FullPageNotFoundView
+            shouldShow
+            onBackButtonPress={() => Navigation.goBack(ROUTES.SETTINGS_WORKSPACES)}
+            shouldForceFullScreen
+        />
+    ) : (
+        <NotFoundPage onBackButtonPress={() => Navigation.goBack(ROUTES.WORKSPACE_PROFILE.getRoute(policyID))} />
+    );
+}
```

Comment:
I think no need for using both `FullPageNotFoundView` and `NotFoundPage` because `NotFoundPage` is the same as `FullPageNotFoundView` else if we pass prop `shouldForceFullScreen = true`.
What do you think about using only `FullPageNotFoundView` like this?
```javascript
<FullPageNotFoundView
    shouldShow
    onBackButtonPress={() => Navigation.goBack(!isFeatureEnabled ? ROUTES.SETTINGS_WORKSPACES : ROUTES.WORKSPACE_PROFILE.getRoute(policyID))}
    shouldForceFullScreen={!isFeatureEnabled}
/>
```

Reply:
We can also use it inline and remove `PageNotFoundFallback` component.

Reply:
`NotFoundPage` is inside of a `ScreenWrapper`, which has a `SafeAreaView` included. If we replace it with `FullPageNotFoundView`, it will break the mobile version of the app as shown below. It's been done this way to preserve the logic present in the previous version. 
I think we should take the wider approach and unify them both in a separate ticket.

<img src="https://github.com/Expensify/App/assets/45288762/515e4a39-1626-46f2-a9cd-f6947f61fd71" width="200" />